File: 2024/03/1.py
# Part 1
import re
with open('input.txt', 'r') as file:
  text = file.read().replace('\n', '')

pattern = r"mul\((-?\d+),(-?\d+)\)"

matches = re.findall(pattern, text)
sum = 0

for x, y in matches:
  sum += int(x)*int(y)

print(sum)

# Part 2
import re
with open('input.txt', 'r') as file:
  raw_text = file.read().replace('\n', '')

do_text = raw_text.split('do()')

for index, t in enumerate(do_text):
  do_text[index] = t.split('don\'t()')[0]

pattern = r"mul\((-?\d+),(-?\d+)\)"

matches = []
for text in do_text:
  if (text == ''): continue
  matches += re.findall(pattern, text)

sum = 0

for x, y in matches:
  sum += int(x)*int(y)

print(sum)

File: 2024/03/1.js
const fs = require("node:fs");

const file = fs.readFileSync("input.txt", "utf8");

const mulRe = /mul\(\d{1,3},\d{1,3}\)/g;

const insRe = /(?<=don't\(\))(.*?)(?=do\(\))/;

const insMatches = file.match(insRe);

console.log(file);
console.log(insMatches);

let fileCopy = file;
for (const insMatch of insMatches) {
  fileCopy = fileCopy.replace(insMatch, "");
}

const mulMatches = fileCopy.match(mulRe);

let ans = 0
for (const mulMatch of mulMatches) {
  ans += mul(mulMatch)
}

console.log("ans", ans);

function mul(str) {
  const [a, b] = str.replace("mul(", "").replace(")", "").split(",");
  return a * b;
}
